File: scripts/ListBoletosPo/listBoletosPo.py
#!/usr/bin/env python3
"""Lista boletos de um processo (ListBoletosPO) no webservice Penhora Online da ONR."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from dotenv import load_dotenv

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from lib.onr_env import env_int  # noqa: E402
from lib.onr_json import to_json_safe  # noqa: E402
from lib.onr_penhora_online import (  # noqa: E402
    hash_error_hint,
    load_login_config,
    load_penhora_online_soap_config,
    load_serventia_chave,
    resolve_auth_hash,
)
from lib.onr_soap import call_operation_from_cfg  # noqa: E402
from lib.onr_zeep_serialize import serialize_zeep_list  # noqa: E402

FIELD_ORDER = ("Hash", "IDProcesso")


def load_config() -> dict:
    load_dotenv(ROOT / ".env")

    id_processo = env_int("PENHORA_ONLINE_ID_PROCESSO")
    if id_processo is None:
        raise SystemExit(
            "Defina PENHORA_ONLINE_ID_PROCESSO no .env com o ID do processo "
            "(obtido em GetPedidoPO, campo IDProcesso)."
        )

    return {
        "chave": load_serventia_chave(),
        "login_cfg": load_login_config(),
        "id_processo": id_processo,
        **load_penhora_online_soap_config(),
    }


def build_request(hash_value: str, id_processo: int) -> dict:
    values = {"Hash": hash_value, "IDProcesso": id_processo}
    return {key: values[key] for key in FIELD_ORDER}


def build_response(result) -> dict:
    return {
        "RETORNO": result.RETORNO,
        "CODIGOERRO": result.CODIGOERRO,
        "ERRODESCRICAO": result.ERRODESCRICAO,
        "Boletos": serialize_zeep_list(result.Boletos, "ListBoletosPO_Boleto_WSResp"),
    }


def main() -> int:
    cfg = load_config()
    hash_value = resolve_auth_hash(cfg["chave"], cfg["login_cfg"])
    o_request = build_request(hash_value, cfg["id_processo"])

    print("=== Parâmetros ListBoletosPO ===")
    print(json.dumps(o_request, ensure_ascii=False, indent=2))
    print(f"\nEndpoint: {cfg['endpoint']}")

    result = call_operation_from_cfg(cfg, "ListBoletosPO", o_request)
    response = to_json_safe(build_response(result))

    print("\n=== Resposta ===")
    print(json.dumps(response, ensure_ascii=False, indent=2))

    if not response.get("RETORNO"):
        codigo = response.get("CODIGOERRO", "?")
        descricao = response.get("ERRODESCRICAO", "")
        print(f"\nListBoletosPO falhou: [{codigo}] {descricao}", file=sys.stderr)
        hint = hash_error_hint(codigo) if isinstance(codigo, int) else None
        if hint:
            print(hint, file=sys.stderr)
        return 1

    boletos_count = len(response.get("Boletos") or [])
    print(f"\nOK — {boletos_count} boleto(s) retornado(s).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/ListBoletosPo/listBoletosPo.js
/**
 * Lista boletos de um processo (ListBoletosPO) no webservice Penhora Online da ONR.
 */
import fs from "node:fs";
import soap from "soap";
import { envInt, resolvePath } from "../../lib/onr_env.js";
import {
  hashErrorHint,
  loadLoginConfig,
  loadPenhoraOnlineSoapConfig,
  loadServentiaChave,
  resolveAuthHash,
} from "../../lib/onr_penhora_online.js";

const FIELD_ORDER = ["Hash", "IDProcesso"];

function loadConfig() {
  const idProcesso = envInt("PENHORA_ONLINE_ID_PROCESSO");
  if (idProcesso === undefined) {
    throw new Error(
      "Defina PENHORA_ONLINE_ID_PROCESSO no .env com o ID do processo " +
        "(obtido em GetPedidoPO, campo IDProcesso)."
    );
  }

  return {
    chave: loadServentiaChave(),
    loginCfg: loadLoginConfig(),
    idProcesso,
    ...loadPenhoraOnlineSoapConfig(),
  };
}

function buildRequest(hashValue, idProcesso) {
  const values = { Hash: hashValue, IDProcesso: idProcesso };
  return Object.fromEntries(FIELD_ORDER.map((key) => [key, values[key]]));
}

function serializeBoletos(boletos) {
  if (!boletos) return [];
  const items = boletos.ListBoletosPO_Boleto_WSResp;
  if (!items) return [];
  return Array.isArray(items) ? items : [items];
}

function normalizeResponse(result) {
  return {
    RETORNO: result.RETORNO,
    CODIGOERRO: result.CODIGOERRO,
    ERRODESCRICAO: result.ERRODESCRICAO,
    Boletos: serializeBoletos(result.Boletos),
  };
}

async function listBoletosPo(cfg, oRequest) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL não encontrado: ${wsdlPath}`);
  }

  const client = await soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 60000 },
  });

  let response;
  try {
    [response] = await client.ListBoletosPOAsync({ oRequest });
  } catch (err) {
    const status = err?.response?.status;
    const body =
      typeof err?.body === "string"
        ? err.body
        : typeof err?.response?.data === "string"
          ? err.response.data
          : null;

    if (status === 503 || (body && !body.trimStart().startsWith("<"))) {
      throw new Error(
        `Servidor ONR indisponível (HTTP ${status ?? "?"}): ${body ?? err.message}`
      );
    }
    throw err;
  }

  const result =
    response?.ListBoletosPOResult ??
    response?.ListBoletosPOResponse?.ListBoletosPOResult ??
    response;

  return normalizeResponse(result);
}

async function main() {
  const cfg = loadConfig();
  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg);
  const oRequest = buildRequest(hash, cfg.idProcesso);

  console.log("=== Parâmetros ListBoletosPO ===");
  console.log(JSON.stringify(oRequest, null, 2));
  console.log(`\nEndpoint: ${cfg.endpoint}`);

  const response = await listBoletosPo(cfg, oRequest);
  console.log("\n=== Resposta ===");
  console.log(JSON.stringify(response, null, 2));

  if (!response.RETORNO) {
    console.error(
      `\nListBoletosPO falhou: [${response.CODIGOERRO}] ${response.ERRODESCRICAO}`
    );
    const hint = hashErrorHint(response.CODIGOERRO);
    if (hint) console.error(hint);
    process.exit(1);
  }

  console.log(`\nOK — ${response.Boletos.length} boleto(s) retornado(s).`);
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
